File: ios/Harakat Lens/Views/QuranMatchView.swift
//
//  QuranMatchView.swift
//  Harakat Lens
//
//  Adapted for the library plugin slot: the audio service is injected as an
//  `EveryAyahAudioProvider` reference (instead of `@Environment` from the
//  app), and the brand-green colour is a local constant (the library's
//  `AppDesign` doesn't vend it). Layout and behaviour are unchanged.
//

import BiangBiangUI
import SwiftUI

struct QuranMatchView: View {
    let match: QuranMatch
    let surahName: SurahName?
    @Bindable var audio: EveryAyahAudioProvider
    /// Shows the ✕ button. Only the inline (text) card needs it; the camera
    /// half-sheet is dismissed by swiping down, so the ✕ would be redundant.
    var showsDismissButton: Bool
    /// User dismissal. The plugin clears the sticky match and starts the
    /// respawn cooldown so the same verse does not immediately reappear.
    var onDismiss: () -> Void

    private let brandGreen = Color(red: 0x00 / 255.0, green: 0x6C / 255.0, blue: 0x35 / 255.0)

    var body: some View {
        VStack(alignment: .leading, spacing: AppDesign.stackSpacing) {
            HStack(alignment: .firstTextBaseline) {
                QuranMatchHeaderView(headerLine: headerLine, brandGreen: brandGreen)
                if showsDismissButton {
                    Spacer()
                    Button {
                        audio.stop()
                        onDismiss()
                    } label: {
                        Image(systemName: "xmark.circle.fill")
                            .foregroundStyle(.secondary)
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel("Dismiss Quran match")
                }
            }
            Text(match.ayah.text)
                .font(.title2)
                .multilineTextAlignment(.trailing)
                .environment(\.layoutDirection, .rightToLeft)
                .frame(maxWidth: .infinity, alignment: .trailing)
            Text(match.ayah.transliteration)
                .font(.body.italic())
                .foregroundStyle(.secondary)
            Text(match.ayah.translationEn)
                .font(.body)
            HStack {
                listenButton
                Button("Copy", systemImage: "doc.on.doc", action: copyToClipboard)
                    .buttonStyle(.bordered)
                ShareLink(item: formattedForCopy) {
                    Label("Share", systemImage: "square.and.arrow.up")
                }
                .buttonStyle(.bordered)
            }
        }
        .padding()
        .background(
            RoundedRectangle(cornerRadius: AppDesign.cornerRadius)
                .fill(brandGreen.opacity(0.08))
        )
        .overlay(alignment: .center) {
            RoundedRectangle(cornerRadius: AppDesign.cornerRadius)
                .stroke(brandGreen.opacity(0.6), lineWidth: 1)
        }
    }

    private var headerLine: String {
        let nameLabel: String = {
            if let s = surahName {
                return "\(s.transliteration) (\(s.english))"
            }
            return "Surah \(match.ayah.surah)"
        }()
        return "Surah \(match.ayah.surah) · \(nameLabel) · Ayah \(match.ayah.ayah)"
    }

    private var formattedForCopy: String {
        """
        \(headerLine)

        \(match.ayah.text)

        \(match.ayah.transliteration)

        \(match.ayah.translationEn)
        """
    }

    private func copyToClipboard() {
        UIPasteboard.general.string = formattedForCopy
    }

    @ViewBuilder
    private var listenButton: some View {
        let surah = match.ayah.surah
        let ayah = match.ayah.ayah
        let isLoading = audio.isLoadingAyah(surah: surah, ayah: ayah)
        let isPlaying = audio.isPlayingAyah(surah: surah, ayah: ayah)
        Button {
            if isPlaying || isLoading {
                audio.stop()
            } else {
                audio.playAyah(surah: surah, ayah: ayah)
            }
        } label: {
            if isLoading {
                HStack(spacing: 6) {
                    ProgressView().controlSize(.small)
                    Text("Loading")
                }
            } else if isPlaying {
                Label("Stop", systemImage: "stop.circle.fill")
            } else {
                Label("Listen", systemImage: "play.circle.fill")
            }
        }
        .buttonStyle(.bordered)
        .accessibilityHint("Stream the recitation of this ayah")
    }
}

private struct QuranMatchHeaderView: View {
    let headerLine: String
    let brandGreen: Color

    var body: some View {
        Label(headerLine, systemImage: "book.closed.fill")
            .font(.subheadline.weight(.semibold))
            .foregroundStyle(brandGreen)
    }
}

File: android/app/src/main/java/dev/veeso/harakatlens/arabic/quran/QuranMatchView.kt
package dev.veeso.harakatlens.arabic.quran

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import dev.veeso.biangbiangui.ui.AppDesign

/**
 * Ported faithfully from the Harakat-Lens iOS app and the already-complete
 * iOS `ArabicExample/Quran/QuranMatchView`. Two adaptations for the library
 * plugin slot (mirroring iOS): the audio service is injected as an
 * [EveryAyahAudioProvider] reference, and the brand-green colour is a local
 * constant (the library's `AppDesign` doesn't vend it). Layout and behaviour
 * follow the SwiftUI view.
 */
/** Brand green `#006C35` (the library's `AppDesign` doesn't vend it). */
private val brandGreen = Color(0xFF006C35)

@Composable
fun QuranMatchView(
    match: QuranMatch,
    surahName: SurahName?,
    audio: EveryAyahAudioProvider,
) {
    val state by audio.state.collectAsState()
    val surah = match.ayah.surah
    val ayah = match.ayah.ayah
    val isLoading = state is EveryAyahAudioProvider.State.LoadingAyah &&
        (state as EveryAyahAudioProvider.State.LoadingAyah).let { it.surah == surah && it.ayah == ayah }
    val isPlaying = state is EveryAyahAudioProvider.State.PlayingAyah &&
        (state as EveryAyahAudioProvider.State.PlayingAyah).let { it.surah == surah && it.ayah == ayah }

    val nameLabel = surahName
        ?.let { "${it.transliteration} (${it.english})" }
        ?: "Surah ${match.ayah.surah}"
    val headerLine = "Surah ${match.ayah.surah} · $nameLabel · Ayah ${match.ayah.ayah}"

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .background(
                color = brandGreen.copy(alpha = 0.08f),
                shape = RoundedCornerShape(AppDesign.cornerRadius),
            )
            .border(
                width = 1.dp,
                color = brandGreen.copy(alpha = 0.6f),
                shape = RoundedCornerShape(AppDesign.cornerRadius),
            )
            .padding(AppDesign.horizontalPadding),
        verticalArrangement = Arrangement.spacedBy(AppDesign.stackSpacing),
    ) {
        Text(text = "📖  $headerLine", color = brandGreen)
        Text(
            text = match.ayah.text,
            modifier = Modifier.fillMaxWidth(),
            textAlign = TextAlign.End,
        )
        Text(
            text = match.ayah.transliteration,
            fontStyle = FontStyle.Italic,
        )
        Text(text = match.ayah.translationEn)
        Row(horizontalArrangement = Arrangement.spacedBy(AppDesign.stackSpacing)) {
            OutlinedButton(onClick = {
                if (isPlaying || isLoading) {
                    audio.stop()
                } else {
                    audio.playAyah(surah, ayah)
                }
            }) {
                Text(
                    when {
                        isLoading -> "Loading"
                        isPlaying -> "Stop"
                        else -> "Listen"
                    },
                )
            }
        }
    }
}